### DISCUSSION TITLE
How would I get the Extended public key as a zpub for a given mnemonic?

### DISCUSSION THREAD

**Discussion by nondiremanuel** [bitcoindevkit/bdk#977] [Q&A]
Repost from issue #973 by @spicefarer:

I've been looking into this for hours now and I can't figure out how to get the zpub for a given mnemonic. I'm using [this website](https://guggero.github.io/cryptography-toolkit/#!/hd-wallet) as a reference.

Here, the given mnemonic (elegant aerobic pony glare science valid tool power toast effort loud bomb) leads to this xpub:

`xpub661MyMwAqRbcFnVHjPTLtR5ZHJkg3iHG4uymkHkTPWveNvLxxjjHyAM5RwPstw1rvC1wk2Ls7b38qp9rriGUSwHPwWZhyBf7HkKV9yekFtB`

and this zpub (when selecting BIP84):

`zpub6jftahH18ngZxNsXQ72bJbGZdF3ZvxGFu92DK5YE9XgQV7yRU44RDHfMUMK3tkKhjUFZEyXz2ukEcPNzJ76W3QebgBxZ91J5qCSmw5AgA1S`

I can successfully get the xpub with bdk, but all my attempts at getting an xpub failed. This is my current code:

```
use std::str::FromStr;

use bdk::{
    bitcoin::secp256k1::Secp256k1,
    keys::{
        bip39::{Language, Mnemonic},
        DerivableKey, ExtendedKey,
    },
    miniscript::Legacy,
};
use bitcoin::{
    util::bip32::{DerivationPath, ExtendedPubKey},
    Network,
};

pub fn derive_zpub_from_mnemonic() -> Result<String, Box<dyn std::error::Error>> {
    let mnemonic_string = "test walk nut penalty hip pave soap entry language right filter choice";
    let mnemonic = Mnemonic::parse_in(Language::English, mnemonic_string)?;
    let network = Network::Bitcoin;
    let secp = Secp256k1::new();

    let extdk: ExtendedKey<Legacy> = mnemonic.into_extended_key()?;
    let xpriv = extdk.into_xprv(network).unwrap();

    // Derive zPriv from xPriv
    let zpriv_derivation_path = "m/84'/0'/0'";
    let path: Vec<_> = DerivationPath::from_str(zpriv_derivation_path)?.into();
    let zpriv = xpriv.derive_priv(&secp, &path)?;

    let zpub = ExtendedPubKey::from_priv(&secp, &zpriv);
    Ok(zpub.to_string())
}
```
Thanks in advance!

**Comment by spicefarer** [comment]
Thanks for reposting @nondiremanuel 

I've found a kind of solution now by using the `bip32` crate:

``` rs
let mnemonic = Mnemonic::parse_in(Language::English, mnemonic_string)?;
let network = Network::Bitcoin;
let secp = Secp256k1::new();

let seed = mnemonic.to_seed("");
let root_xprv = bip32::XPrv::new(&seed)?;
let child_xpub = root_xprv.public_key();
let zpub = child_xpub.to_string(Prefix::ZPUB);
```

This works / gives me a zpub, however the zpub doesn't resolve to the addresses used in my wallet (when checking for balances), and, it is a different zpub from the one that the Electrum wallet shows. So my solution is still wrong somehow and I don't understand why. It would be great if BDK could offer a way to replicate getting zpubs, ypubs, tpubs similar to what Electrum / other wallets are doing. (my use case is using a third party API to find all balances for a given mnemonic in one go)

**Comment by thunderbiscuit** [comment]
Hi Spicefarer. Take a look at [this message](https://discord.com/channels/753336465005608961/759431700807286795/1088227235581927434) and the following exchange concerning z/y/v pubs.

I think there may or may not be separate crates (probably in various states of "not-so-maintained"? not sure... and I certainly don't recommend them) that add it, but the BDK approach follows rust-bitcoin and doesn't use z/y/v pubs. Using output descriptors removes the requirement for those with much more upside.
